File: automation_cron.py
import json
import os
import sys
from datetime import datetime, timedelta

DB_PATH = os.path.join(os.path.dirname(__file__), 'db.json')

def load_db():
    if not os.path.exists(DB_PATH):
        # Initial database state
        initial_data = {
            "goals": [
                {
                    "id": 101,
                    "name": "Cuenta de Ahorro",
                    "targetAmount": 100000.0,
                    "currentAmount": 100000.0,  # Initial balance 100,000 according to requirement
                    "currency": "USD",
                    "transactions": [],
                    "createdAt": datetime.now().isoformat()
                }
            ],
            "automations": [
                {
                    "id": "rule_1",
                    "goalId": 101,
                    "actionType": "remove",  # Retiro (-)
                    "amount": 28.0,  # Discount of 28 daily
                    "frequency": "daily",
                    "note": "Agua",
                    "lastRun": None,
                    "nextRun": (datetime.now().replace(hour=1, minute=0, second=5, microsecond=0)).isoformat(),
                    "createdAt": datetime.now().isoformat()
                }
            ]
        }
        with open(DB_PATH, 'w', encoding='utf-8') as f:
            json.dump(initial_data, f, indent=2, ensure_ascii=False)
        print(f"[DB] Base de datos inicial creada en {DB_PATH}")

    with open(DB_PATH, 'r', encoding='utf-8') as f:
        return json.load(f)

def save_db(data):
    with open(DB_PATH, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

def calculate_next_run_date(from_date_str, frequency):
    from_date = datetime.fromisoformat(from_date_str)
    # Ensure it's 1:00:05 AM
    from_date = from_date.replace(hour=1, minute=0, second=5, microsecond=0)
    
    if frequency == 'daily':
        next_date = from_date + timedelta(days=1)
    elif frequency == 'weekly':
        next_date = from_date + timedelta(weeks=1)
    elif frequency == 'monthly':
        # Simple month increment
        month = from_date.month + 1
        year = from_date.year
        if month > 12:
            month = 1
            year += 1
        # Handle end-of-month days safely
        day = min(from_date.day, 28)
        next_date = from_date.replace(year=year, month=month, day=day)
    elif frequency == 'yearly':
        next_date = from_date.replace(year=from_date.year + 1)
    else:
        next_date = from_date + timedelta(days=1)
        
    return next_date.isoformat()

def run_automations():
    db = load_db()
    now = datetime.now()
    updated = False

    print(f"\n[{now.strftime('%Y-%m-%d %H:%M:%S')}] Procesando reglas de automatización...")

    for rule in db["automations"]:
        next_run_date = datetime.fromisoformat(rule["nextRun"])
        goal = next((g for g in db["goals"] if g["id"] == int(rule["goalId"])), None)

        if not goal:
            print(f"Warning: Meta vinculada {rule['goalId']} no encontrada para la regla '{rule['note']}'.")
            continue

        while now >= next_run_date:
            amount = rule["amount"]
            action_text = "Ingreso automático" if rule["actionType"] == "add" else "Retiro automático"
            old_amount = goal["currentAmount"]

            # Alter balance only on that account (Aislamiento de Cuentas)
            if rule["actionType"] == "add":
                goal["currentAmount"] += amount
            else:
                goal["currentAmount"] = max(0.0, goal["currentAmount"] - amount)

            transaction = {
                "id": datetime.now().timestamp() + 1.1,
                "amount": amount if rule["actionType"] == "add" else -amount,
                "note": f"{action_text} - Nota: {rule['note']}",
                "date": next_run_date.isoformat(),
                "type": rule["actionType"],
                "automationId": rule["id"]
            }

            if "transactions" not in goal or not goal["transactions"]:
                goal["transactions"] = []
            goal["transactions"].insert(0, transaction)

            print(f"Executed rule '{rule['note']}' ({'+' if rule['actionType'] == 'add' else '-'}{amount}) en '{goal['name']}':")
            print(f"   Saldo anterior: {old_amount} -> Nuevo saldo: {goal['currentAmount']}")

            rule["lastRun"] = next_run_date.isoformat()
            next_run_date = datetime.fromisoformat(calculate_next_run_date(next_run_date.isoformat(), rule["frequency"]))
            rule["nextRun"] = next_run_date.isoformat()
            updated = True

    if updated:
        save_db(db)
        print("[DB] Base de datos actualizada y guardada correctamente.")
    else:
        print("[INFO] No hay ejecuciones pendientes para procesar.")

def edit_automation_note(rule_id, new_note, option):
    db = load_db()
    rule = next((r for r in db["automations"] if r["id"] == rule_id), None)

    if not rule:
        print(f"Error: No se encontró la regla con ID '{rule_id}'")
        return

    old_note = rule["note"]
    rule["note"] = new_note

    print(f"\nEditing rule '{rule_id}': Cambiando nota de \"{old_note}\" a \"{new_note}\"")

    if option in ('B', 'b'):
        tx_updated = 0
        for goal in db["goals"]:
            if "transactions" in goal:
                for tx in goal["transactions"]:
                    old_action_text_add = f"Ingreso automático - Nota: {old_note}"
                    old_action_text_remove = f"Retiro automático - Nota: {old_note}"

                    matches_id = tx.get("automationId") == rule["id"]
                    matches_note = tx["note"] in (old_action_text_add, old_action_text_remove)

                    if matches_id or matches_note:
                        action_text = "Ingreso automático" if tx["amount"] > 0 else "Retiro automático"
                        tx["note"] = f"{action_text} - Nota: {new_note}"
                        tx["automationId"] = rule["id"]
                        tx_updated += 1
        print(f"Option B selected: Se actualizaron {tx_updated} transacciones en el historial pasado.")
    else:
        print(f"Option A selected: El historial anterior se mantiene intacto. Solo los nuevos movimientos usarán \"{new_note}\".")

    save_db(db)
    print("[DB] Base de datos guardada.")

def show_status():
    db = load_db()
    print("\n===== ESTADO DE LA BASE DE DATOS =====")
    print(f"\nCuentas/Metas ({len(db['goals'])}):")
    for goal in db["goals"]:
        print(f"- [ID: {goal['id']}] {goal['name']}: Saldo actual = {goal['currentAmount']} {goal['currency']}")
        txs = goal.get("transactions", [])
        print(f"  Historial de Transacciones ({len(txs)}):")
        for tx in txs[:5]:
            print(f"    * [{tx['date']}] { '+' if tx['amount'] > 0 else ''}{tx['amount']} | {tx['note']}")
        if len(txs) > 5:
            print(f"    * ... y {len(txs) - 5} transacciones más")

    print(f"\nReglas de Automatización ({len(db['automations'])}):")
    for rule in db["automations"]:
        goal = next((g for g in db["goals"] if g["id"] == int(rule["goalId"])), None)
        print(f"- [ID: {rule['id']}] Nota/Concepto: \"{rule['note']}\"")
        print(f"  Vínculo: {goal['name'] if goal else 'Meta Desconocida'} ({'Ingreso' if rule['actionType'] == 'add' else 'Retiro'})")
        print(f"  Detalles: Monto = {rule['amount']} | Frecuencia = {rule['frequency']}")
        print(f"  Próxima corrida: {rule['nextRun']}")
    print("=======================================")

if __name__ == '__main__':
    if len(sys.argv) > 1:
        command = sys.argv[1]
        if command == '--run':
            run_automations()
        elif command == '--edit':
            if len(sys.argv) < 5:
                print("Error: Faltan parámetros para --edit. Uso: python automation_cron.py --edit <id> <nueva_nota> <A|B>")
            else:
                edit_automation_note(sys.argv[2], sys.argv[3], sys.argv[4])
        elif command == '--status':
            show_status()
        else:
            print("Comandos disponibles: --run, --status, --edit <id> <nueva_nota> <A|B>")
    else:
        print("\n=== MiAhorro Automatizaciones Backend Python Cronjob ===")
        run_automations()
        show_status()

File: automation_cron.js
/**
 * ===== SCRIPT BACKEND CRONJOB DE AUTOMATIZACIÓN =====
 * Este script procesa las reglas de automatización configuradas.
 * Puede ser ejecutado de forma programada a la 1:00 AM todos los días.
 * 
 * Uso: node automation_cron.js [opciones]
 * Opciones:
 *   --run             Ejecuta el procesamiento de automatizaciones pendientes (comportamiento cronjob regular)
 *   --edit <id> <nueva_nota> <opcion_A_o_B>
 *                     Simula la edición de una nota de regla y aplica la opción A o B al historial.
 *   --status          Muestra el estado actual de las cuentas (metas) y reglas.
 */

const fs = require('fs');
const path = require('path');

const DB_PATH = path.join(__dirname, 'db.json');

// Cargar Base de Datos Local
function loadDB() {
    if (!fs.existsSync(DB_PATH)) {
        // Inicializar base de datos vacía si no existe
        const initialData = {
            goals: [
                {
                    id: 101,
                    name: "Cuenta de Ahorro",
                    targetAmount: 100000,
                    currentAmount: 100000, // Saldo inicial 100,000 según el ejemplo
                    currency: "USD",
                    transactions: [],
                    createdAt: new Date().toISOString()
                }
            ],
            automations: [
                {
                    id: "rule_1",
                    goalId: 101,
                    actionType: "remove", // Retiro (-)
                    amount: 28, // Descuento de 28 diario según el ejemplo
                    frequency: "daily",
                    note: "Agua",
                    lastRun: null,
                    nextRun: getTodayAtOneAM().toISOString(),
                    createdAt: new Date().toISOString()
                }
            ]
        };
        fs.writeFileSync(DB_PATH, JSON.stringify(initialData, null, 2), 'utf8');
        console.log(`📁 Base de datos inicial creada en ${DB_PATH}`);
    }
    
    return JSON.parse(fs.readFileSync(DB_PATH, 'utf8'));
}

function saveDB(data) {
    fs.writeFileSync(DB_PATH, JSON.stringify(data, null, 2), 'utf8');
}

function getTodayAtOneAM() {
    const date = new Date();
    date.setHours(1, 0, 5, 0); // 1:00 AM (justo después)
    return date;
}

// Fechas y cálculos de frecuencia
function calculateNextRunDate(fromDate, frequency) {
    const date = new Date(fromDate);
    date.setHours(1, 0, 5, 0);
    
    switch (frequency) {
        case 'daily':
            date.setDate(date.getDate() + 1);
            break;
        case 'weekly':
            date.setDate(date.getDate() + 7);
            break;
        case 'monthly':
            date.setMonth(date.getMonth() + 1);
            break;
        case 'yearly':
            date.setFullYear(date.getFullYear() + 1);
            break;
    }
    return date.toISOString();
}

// 1. Ejecutar Automatizaciones
function runAutomations() {
    const db = loadDB();
    const now = new Date();
    let updated = false;

    console.log(`\n⏰ [${now.toLocaleString()}] Procesando reglas de automatización...`);

    db.automations.forEach(rule => {
        let nextRunDate = new Date(rule.nextRun);
        const goal = db.goals.find(g => g.id === Number(rule.goalId));

        if (!goal) {
            console.log(`⚠️ Meta vinculada ${rule.goalId} no encontrada para la regla '${rule.note}'.`);
            return;
        }

        // Ejecutar ciclos pendientes transcurridos
        while (now >= nextRunDate) {
            const amount = rule.amount;
            const actionText = rule.actionType === 'add' ? 'Ingreso automático' : 'Retiro automático';
            const oldAmount = goal.currentAmount;
            
            // Alterar saldo SOLO en esa cuenta (Aislamiento de Cuentas)
            if (rule.actionType === 'add') {
                goal.currentAmount += amount;
            } else {
                goal.currentAmount = Math.max(0, goal.currentAmount - amount);
            }
            
            const transaction = {
                id: Date.now() + Math.random(),
                amount: rule.actionType === 'add' ? amount : -amount,
                note: `${actionText} - Nota: ${rule.note}`,
                date: nextRunDate.toISOString(),
                type: rule.actionType,
                automationId: rule.id
            };
            
            if (!goal.transactions) goal.transactions = [];
            goal.transactions.unshift(transaction);

            console.log(`⚡ Ejecutada regla '${rule.note}' (${rule.actionType === 'add' ? '+' : '-'}${amount}) en '${goal.name}':`);
            console.log(`   Saldo anterior: ${oldAmount} -> Nuevo saldo: ${goal.currentAmount}`);
            
            // Actualizar tiempos
            rule.lastRun = nextRunDate.toISOString();
            nextRunDate = new Date(calculateNextRunDate(nextRunDate, rule.frequency));
            rule.nextRun = nextRunDate.toISOString();
            updated = true;
        }
    });

    if (updated) {
        saveDB(db);
        console.log(`✅ Base de datos actualizada y guardada correctamente.`);
    } else {
        console.log(`ℹ️ No hay ejecuciones pendientes para procesar.`);
    }
}

// 2. Editar Nota de Automatización (Simulación de Opción A/B)
function editAutomationNote(ruleId, newNote, option) {
    const db = loadDB();
    const rule = db.automations.find(r => r.id === ruleId);

    if (!rule) {
        console.log(`❌ Error: No se encontró la regla con ID '${ruleId}'`);
        return;
    }

    const oldNote = rule.note;
    rule.note = newNote; // Actualizar nota en la regla

    console.log(`\n✏️ Editando regla '${ruleId}': Cambiando nota de "${oldNote}" a "${newNote}"`);

    if (option === 'B' || option === 'b') {
        // Opción B: Actualizar también historial pasado
        let txUpdated = 0;
        db.goals.forEach(goal => {
            if (goal.transactions) {
                goal.transactions.forEach(tx => {
                    // Coincide por automationId o por formato de nota anterior
                    const oldActionTextAdd = `Ingreso automático - Nota: ${oldNote}`;
                    const oldActionTextRemove = `Retiro automático - Nota: ${oldNote}`;
                    
                    const matchesId = tx.automationId === rule.id;
                    const matchesNote = tx.note === oldActionTextAdd || tx.note === oldActionTextRemove;

                    if (matchesId || matchesNote) {
                        const actionText = tx.amount > 0 ? 'Ingreso automático' : 'Retiro automático';
                        tx.note = `${actionText} - Nota: ${newNote}`;
                        tx.automationId = rule.id; // Vincular explícitamente
                        txUpdated++;
                    }
                });
            }
        });
        console.log(`✅ Opción B seleccionada: Se actualizaron ${txUpdated} transacciones en el historial pasado.`);
    } else {
        // Opción A: Mantener el historial intacto
        console.log(`ℹ️ Opción A seleccionada: El historial anterior se mantiene intacto. Solo los nuevos movimientos automáticos usarán "${newNote}".`);
    }

    saveDB(db);
    console.log(`💾 Base de datos guardada.`);
}

// 3. Mostrar Estado
function showStatus() {
    const db = loadDB();
    console.log(`\n===== ESTADO DE LA BASE DE DATOS =====`);
    
    console.log(`\nCuentas/Metas (${db.goals.length}):`);
    db.goals.forEach(goal => {
        console.log(`- [ID: ${goal.id}] ${goal.name}: Saldo actual = ${goal.currentAmount} ${goal.currency}`);
        console.log(`  Historial de Transacciones (${goal.transactions ? goal.transactions.length : 0}):`);
        if (goal.transactions && goal.transactions.length > 0) {
            goal.transactions.slice(0, 5).forEach(tx => {
                console.log(`    * [${new Date(tx.date).toLocaleDateString()} ${new Date(tx.date).toLocaleTimeString()}] ${tx.amount > 0 ? '+' : ''}${tx.amount} | ${tx.note}`);
            });
            if (goal.transactions.length > 5) console.log(`    * ... y ${goal.transactions.length - 5} transacciones más`);
        } else {
            console.log(`    * Sin transacciones`);
        }
    });

    console.log(`\nReglas de Automatización (${db.automations.length}):`);
    db.automations.forEach(rule => {
        const goal = db.goals.find(g => g.id === Number(rule.goalId));
        console.log(`- [ID: ${rule.id}] Nota/Concepto: "${rule.note}"`);
        console.log(`  Vínculo: ${goal ? goal.name : 'Meta Desconocida'} (${rule.actionType === 'add' ? 'Ingreso' : 'Retiro'})`);
        console.log(`  Detalles: Monto = ${rule.amount} | Frecuencia = ${rule.frequency}`);
        console.log(`  Próxima corrida: ${new Date(rule.nextRun).toLocaleString()}`);
    });
    console.log(`=======================================`);
}

// Parsear Argumentos de CLI
const args = process.argv.slice(2);
const command = args[0];

if (command === '--run') {
    runAutomations();
} else if (command === '--edit') {
    const ruleId = args[1];
    const newNote = args[2];
    const option = args[3]; // 'A' o 'B'

    if (!ruleId || !newNote || !option) {
        console.log("❌ Error: Faltan parámetros para --edit. Uso: node automation_cron.js --edit <id> <nueva_nota> <A|B>");
    } else {
        editAutomationNote(ruleId, newNote, option);
    }
} else if (command === '--status') {
    showStatus();
} else {
    // Si se corre sin argumentos, por defecto inicializamos la BD y ejecutamos
    console.log(`\n=== MiAhorro Automatizaciones Backend Cronjob ===`);
    console.log(`Uso:`);
    console.log(`  node automation_cron.js --run            Procesa automatizaciones pendientes`);
    console.log(`  node automation_cron.js --status         Muestra el estado actual`);
    console.log(`  node automation_cron.js --edit <id> <nueva_nota> <A|B>    Edita nota con opción A o B`);
    console.log(`\nCorriendo por defecto inicialización y ejecución regular...`);
    runAutomations();
    showStatus();
}
